crdt/sequence.py

```diff
@@ -0,0 +1,247 @@
+
+        # Get the new operations to be applied
+        patch_ops = other.operations.get().difference(self.operations.get())
+        
+        # Get a sorted view of the patches to apply
+        patch_log = sorted(patch_ops, key=cmp_to_key(self.compare_operations))
+
+        # Path the sequence using the new operations
+        for op in patch_log:
+            op.do(self.sequence)
+
+        # Merge the two operation logs
+        self.operations = self.operations.merge(other.operations)
+        return self
+
+    def get(self):
+        """
+        Returns the sorted list of payloads in the sequence.
+        """
+        return [obj.operation.payload for obj in self.sequence if not obj.tombstone]
+
+class Object():
+    """
+    An Object represents a single item in a Sequence.
+    """
+    
+    def __init__(self, operation):
+        self.operation = operation
+        self.tombstone = False
+
+    def print(self):
+        """
+        Prints the object to stdout.
+        """
+        print("owner: {}, payload: {}, tombstone: {}".format(self.operation.owner.print(), self.operation.payload, self.tombstone))
+
+class OpId():
+    """
+    An OpId represents a unique identifier for an Operation.
+    """
+
+    def __init__(self, name, id):
+        self.name = name
+        self.id = id
+
+    def compare(self, other):
+        if self.id < other.id:
+            return -1
+        elif self.id > other.id:
+            return 1
+        else:
+            # Conflicts are resolved arbitrarily by comparing the owner IDs
+            # TODO: We could handle "forks" of the operation log and allow for multiple
+            # histories to be valid.
+            if self.name < other.name:
+                return -1
+            elif self.name > other.name:
+                return 1
+            else:
+                return 0
+
+    def is_earlier(self, other):
+        """
+        Returns true if this OpId is earlier than the other OpId.
+        """
+        return self.compare(other) < 0
+
+    def __eq__(self, other):
+        if not isinstance(other, OpId):
+            return False
+        return self.name == other.name and self.id == other.id
+
+    def print(self):
+        """
+        Prints the OpId to stdout.
+        """
+        return "name: {}, id: {}".format(self.name, self.id)
+
+class OperationType(Enum):
+    """
+    Enum for operations.
+    """
+    INSERT = 0
+    REMOVE = 1
+
+class Operation():
+    """
+    An Operation represents a point-in-time change to a Sequence. Every Operation
+    has a globally-unique ID and references a previous Operation, enabling an order of
+    Operations to be imposed.
+    """
+
+    def __init__(self, owner, action, target, payload=None):
+        if action not in OperationType:
+            raise ValueError("Invalid operation type")
+        self.owner = owner
+        self.action = action
+        self.target = target
+        self.payload = payload
```

Comment:
Let me know if the terminology is confusing, the `owner` is the originator of the operation which is an `OpId`. The target is also an `OpId` which is the operation that this operation is performed on.

Reply:
Why don't we go ahead and add this to the docstring to make sure we remember what type each param is and how they're intended to be used. Also, would you mind making them all keyword arguments rather than positional; perhaps they can all default to `None`?